Decouple checking ddl from classifying h2 stmts

escherize opened the pull request:
- This should enable follow-up for easily blocking more types of queries.
- Check *all* statements to ensure they aren't "ddl", not just the 1st one.

Previously, we weren't thoroughly checking every single statement. So if more statements were passed in and the first statement was legit, we allowed them to run. Although H2 shouldn't be used for anything serious, it is still important to plug this vulnerability.

This separation also uncouples the JVM interop details from the purely functional computation of "are any of these statements DDL?". I believe this separation will [simplify detecting or denying query execution](https://github.com/metabase/metabase/pull/28510/files#diff-c4bd5c99a5e021fc7f1ecf5305075b8195b50845e23343846284a821fb0e4a3cR173-R174) based on their [org.h2.command.CommandInterface value](https://github.com/h2database/h2database/blob/master/h2/src/main/org/h2/command/CommandInterface.java#L16-L557).

escherize committed:
Decouple checking ddl from classifying h2 stmts

- should enable followup for easily blocking more kinds of queries
- check all statements to make sure they aren't "ddl".

escherize committed:
fix classify-query

escherize committed:
linter fixes + get-field refactor

escherize commented on src/metabase/driver/h2.clj:
```diff
@@ -127,22 +132,50 @@
+  (< cmd-type CommandInterface/ALTER_SEQUENCE))
+
+(defn- contains-ddl? [{:keys [command-class+types remaining-sql]}]
+  (let [cmd-type-nums (mapv second command-class+types)]
```
second value in each `class+type` here will correspond to a value in https://github.com/h2database/h2database/blob/master/h2/src/main/org/h2/command/CommandInterface.java#L16-L557.

escherize commented on src/metabase/driver/h2.clj:
```diff
@@ -127,22 +132,50 @@
+  (let [cmd-type-nums (mapv second command-class+types)]
+    (boolean
+     (or (some cmd-type-ddl? cmd-type-nums)
+         (some? remaining-sql)))))
```
remaining-sql cannot be checked. so we just assume it contians ddl statements.

dpsutton replied:
if there is any remaining-sql, it means the last parsed sql creates things, and the remaining sql cannot be parsed until the db knows what the new thing is that the remaining sql represents. For instance, the sql

```sql
"CREATE ALIAS MY_SQRT for 'java.lang.Math.sqrt'; select SQRT(2.0);"
```
Results in a CommandList who's first command is `org.h2.command.ddl.CreateFunctionAlias`, has no other prepared commands, and who's remaining sql is `"select SQRT(2.0);"`. It needs the alias to be created before the parser can parse the function call `SQRT`.

Because of this, the presence of `remaining-sql` necessarily means something created something, and we can reject that sql.

escherize replied:
Right, a query statement would have created something _iff it ran_. So, for ddl detection, it means we can reject it. 👍

escherize committed:
return the CommandInterface values as ints

Reach into the CommandList when needed

escherize committed:
docstring wording

escherize committed:
catch invalid queries -- they can't be classified

calherries commented on src/metabase/driver/h2.clj:
```diff
@@ -127,22 +133,51 @@
+(mu/defn ^:private classify-query :- [:map
+                                      [:command-types [:vector pos-int?]]
+                                      [:remaining-sql [:maybe :string]]]
+  "Takes an h2 db id, and a query, returns the command-classes+types (see: ), and any remaining sql.
```
`(see: )`?

escherize replied:
Fixed

calherries reviewed (approved):
Looks tight to me

calherries commented on src/metabase/driver/h2.clj:
```diff
@@ -127,22 +133,51 @@
+
+(defn- check-disallow-ddl-commands [{:keys [database] {:keys [query]} :native :as in}]
+  (let [query-classification (classify-query database query)]
+    (when (and query (contains-ddl? query-classification))
```
`check-disallow-ddl-commands` wouldn't throw an error if `query` is nil before. Now it does, since `(classify-query nil)` raises an exception. 

So, either confirm `query` can never be nil when this function is called and replace the `query` check
```suggestion
    (when (contains-ddl? query-classification)
```
Or wrap the whole block in a `query` check, to preserve behaviour when `query` is nil
```
    (when query ... )
```

escherize replied:
Nice catch, thanks. Done

dpsutton commented on src/metabase/driver/h2.clj:
```diff
@@ -127,22 +133,51 @@
-  (when (and query (contains-ddl? database query))
-    (throw (IllegalArgumentException. "DDL commands are not allowed to be used with h2."))))
+      (let [command-list      (.prepareCommand h2-parser query)
+            command-list-type (.getCommandType ^org.h2.command.CommandList command-list)
```
this typehint is wrong. `prepareCommand` returns a `org.h2.command.Command`. It can be either a `CommandContainer` or a `CommandList`. annotating with the list here is wrong in cases of a single statement in the sql.

But it's also not needed. If you remove it, the parser has type info, and Clojure infers the type of `command-list` from `prepareCommand`.

calherries replied:
It should probably be renamed from `command-list` to `command` too.

escherize replied:
Done

escherize committed:
respond to review comments

- fix see: comment
- remove CommandList typehint
- rename command-list -> command
- skip checking disallowed ddl commands when query is `nil`

then also:
- fix a linter error
- add a test for `nil` query ddl command checking

calherries commented on src/metabase/driver/h2.clj:
```diff
@@ -127,22 +133,54 @@
-      ;; if the query is invalid, then it isn't DDL
-      (catch Throwable _ false))))
+      (let [command           (.prepareCommand h2-parser query)
+            command-list-type (.getCommandType command)
```
```suggestion
            first-command-type (.getCommandType command)
```

calherries reviewed (approved):
Looks tight now

escherize committed:
rename command-list-type -> first-command-type

escherize committed:
fix schema + handle h2-parser build failure

calherries commented on src/metabase/driver/h2.clj:
```diff
@@ -127,22 +133,57 @@
+        {:command-types command-types
+         ;; when there is no remaining sql: return nil for remaining-sql
+         :remaining-sql (get-field command "remaining" nil)})
+      ;; only valid queries can be classified.
+      (catch org.h2.message.DbException _
+        {:command-types [] :remaining-sql nil}))))
```
This means if a query has multiple statements and there's a syntax error in the first that means it fails to be parsed, it could mean following statements get executed, depending on whether the statements are executed independently or not. e.g. `statement with syntax error; CREATE ALIAS ...` will pass the DDL check, but the query will continue to be executed. Luckily the execution of the query is short circuited on the first syntax error and the following statements won't be executed, but that depends on how the query is executed and I'm not sure it's guaranteed to be the case. When I use the H2 web UI, a query like `statement with syntax error; CREATE ALIAS ...` will error on the first statement, and still execute the second statement.

dpsutton replied:
Interesting. That's a good thing to check. But it doesn't seem to be how the driver works in our app.

```clojure
h2=> (metabase.query-processor/process-query
       {:database 1
        :type :native
        :native {:query "select foo from users; CREATE ALIAS MY_SQRT for 'java.lang.Math.sqrt';"}})
Execution error (JdbcSQLSyntaxErrorException) at org.h2.message.DbException/getJdbcSQLException (DbException.java:502).
Table "USERS" not found; SQL statement:
-- Metabase
select foo from users; CREATE ALIAS MY_SQRT for 'java.lang.Math.sqrt'; [42102-212]
```

Here an invalid query followed by a create alias. Our error is only about the syntax error in the first query and nothing about admin requirements in the second.

Similarly, just using the database without our ddl checks and such:

```clojure
h2=> (clojure.java.jdbc/execute!
       (metabase.driver.sql-jdbc.connection/db->pooled-connection-spec 1)
       ["select foo from users; CREATE ALIAS MY_SQRT for 'java.lang.Math.sqrt';"])
Execution error (JdbcSQLSyntaxErrorException) at org.h2.message.DbException/getJdbcSQLException (DbException.java:502).
Table "USERS" not found; SQL statement:
select foo from users; CREATE ALIAS MY_SQRT for 'java.lang.Math.sqrt'; [42102-212]
h2=> (clojure.java.jdbc/execute!
       (metabase.driver.sql-jdbc.connection/db->pooled-connection-spec 1)
       ["select MY_SQRT(4)"])
Execution error (JdbcSQLSyntaxErrorException) at org.h2.message.DbException/getJdbcSQLException (DbException.java:632).
Function "MY_SQRT" not found; SQL statement:
select MY_SQRT(4) [90022-212]
```

Trying to define the function and then use it shows it is not defined.
